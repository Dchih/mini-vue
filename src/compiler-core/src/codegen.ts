import { NodeTypes } from "./ast";
import { TO_DISPLAY_STRING, helperNameMap } from "./runtimeHelpers";

export function generate(ast) {
  const context = createCodegenContext();
  const { push } = context;

  generateFunctionPreamble(ast, context);

  const functionName = "render";
  const args = ["_ctx", "_cache"];
  const signature = args.join(", ");
  push(`return function ${functionName}(${signature}) {`);

  push(`return `);
  genNode(ast.codegen, context);
  push(`}`);
  return {
    code: context.code,
  };

  function generateFunctionPreamble(ast, context) {
    const { push, helper } = context;
    const VueBinging = "Vue";
    const aliasHelper = (s) => `${helper(s)}:_${helper(s)}`;
    if (ast.helpers.length > 0) {
      push(
        `const { ${ast.helpers.map(aliasHelper).join(", ")} } = ${VueBinging}`
      );
    }
    push("\n");
  }
}

function genNode(node, context) {
  switch (node.type) {
    case NodeTypes.TEXT:
      genText(node, context);
      break;
    case NodeTypes.INTERPOLATION:
      genInterpolation(node, context);
      break;
    case NodeTypes.SIMPLE_EXPRESSION:
      genExpression(node, context);
    default:
      break;
  }
}

function createCodegenContext() {
  const context = {
    code: "",
    push(source) {
      context.code += source;
    },
    helper(key) {
      return helperNameMap[key];
    },
  };
  return context;
}

function genText(node, context) {
  const { push } = context;
  push(`'${node.content}'`);
}

function genInterpolation(node, context) {
  const { push, helper } = context;
  push(`_${helper(TO_DISPLAY_STRING)}(`);
  genNode(node.content, context);
  push(`)`);
}
function genExpression(node: any, context: any) {
  const { push } = context;
  push(`${node.content}`);
}
