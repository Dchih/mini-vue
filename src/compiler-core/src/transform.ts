import { NodeTypes } from "./ast";
import { TO_DISPLAY_STRING } from "./runtimeHelpers";

export function transform(root, options = {}) {
  const context = createTransformContext(root, options);
  traverseNode(root, context);
  createRootCodegen(root);
  root.helpers = context.helpers;
}

function createRootCodegen(ast) {
  ast.codegen = ast.children[0];
}

function traverseNode(node: any, context) {
  const { nodeTransforms } = context;
  for (let i = 0; i < nodeTransforms.length; i++) {
    const transform = nodeTransforms[i];
    transform(node);
  }

  switch (node.type) {
    case NodeTypes.INTERPOLATION:
      context.pushHelper(TO_DISPLAY_STRING);
      break;
    case NodeTypes.ROOT:
    case NodeTypes.ELEMENT:
      traverseChildren(context, node);
      break;
    default:
      break;
  }
}

function traverseChildren(context, node) {
  const children = node.children;
  for (let i = 0; i < children.length; i++) {
    const childNode = children[i];
    traverseNode(childNode, context);
  }
}

function createTransformContext(root: any, options: any) {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || [],
    helpers: [] as any[],
    pushHelper(helper: any) {
      context.helpers.push(helper);
    },
  };
  return context;
}
