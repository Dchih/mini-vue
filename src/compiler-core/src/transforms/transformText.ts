import { NodeTypes } from "../ast";

function isText(node) {
  return node.type === NodeTypes.TEXT || node.type === NodeTypes.INTERPOLATION;
}

// 处理 ELEMENT, 将其子节点转换为一个夹带 " + " 的数组
// 故 INTERPOLATION 被覆盖在此节点中
export function transformText(node, context) {
  if (node.type === NodeTypes.ELEMENT) {
    return () => {
      const children = node.children;
      let currentContainer;
      for (let i = 0; i < children.length; i++) {
        const child = children[i];
        if (isText(child)) {
          for (let j = i + 1; j < children.length; j++) {
            const nextChild = children[j];
            if (isText(nextChild)) {
              if (!currentContainer) {
                currentContainer = children[i] = {
                  type: NodeTypes.COMPOUND_EXPRESSION,
                  children: [child],
                };
              }
              currentContainer.children.push(" + ");
              currentContainer.children.push(nextChild);
              children.splice(j, 1);
              j--;
            } else {
              currentContainer = undefined;
              break;
            }
          }
        }
      }
    };
  }
}
