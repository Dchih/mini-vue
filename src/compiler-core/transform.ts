export function transform(root, options) {
  const context = createTransformContext(root, options);
  traverseNode(root, context);
}
function traverseNode(node: any, context) {
  const { nodeTransform } = context;
  for (let i = 0; i < nodeTransform.length; i++) {
    const transform = nodeTransform[i];
    transform(node);
  }
  traverseChildren(context, node);
}

function traverseChildren(context, node) {
  const children = node.children;
  if (children) {
    for (let i = 0; i < children.length; i++) {
      const childNode = children[i];
      traverseNode(childNode, context);
    }
  }
}

function createTransformContext(root: any, options: any) {
  return {
    root,
    nodeTransform: options.nodeTransform || [],
  };
}
