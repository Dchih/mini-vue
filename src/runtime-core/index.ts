export { h } from "./h";
export { renderSlots } from "./helper/renderSlots";
export { createTextNode } from "./vnode";
export { getCurrentInstance } from "./component";
export { provide, inject } from "./injectApi";
export { createRenderer } from "./renderer";
