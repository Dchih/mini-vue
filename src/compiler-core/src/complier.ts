import { generate } from "./codegen";
import { baseParser } from "./parser";
import { transform } from "./transform";
import { transformElement } from "./transforms/transformElement";
import { transformExpression } from "./transforms/transformExpression";
import { transformText } from "./transforms/transformText";

export function baseCompiler(template) {
  const ast = baseParser(template);
  transform(ast, {
    nodeTransforms: [transformExpression, transformElement, transformText],
  });
  const { code } = generate(ast);
  return code;
}
