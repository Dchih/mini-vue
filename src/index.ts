export * from "./runtime-core";
export * from "./runtime-dom";
export * from "./reactivity";

import * as runtimeCore from "./runtime-core";
import { baseCompiler } from "./compiler-core/src";
import { registerFunctionCompiler } from "./runtime-core";
export function compileToFunction(template) {
  const code = baseCompiler(template);
  const render = new Function("Vue", code)(runtimeCore);
  return render;
}

registerFunctionCompiler(compileToFunction);
