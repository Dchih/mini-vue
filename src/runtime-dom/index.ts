import { createRenderer } from "../runtime-core/renderer";

function createElement(type) {
  return document.createElement(type);
}

function patchProp(el, key, val) {
  const isOn = (key) => /^on[A-Z]/.test(key);
  const getEventName = (key) => key.slice(2).toLowerCase();
  if (isOn(key)) {
    el.addEventListener(getEventName(key), val);
  } else {
    el.setAttribute(key, val);
  }
}

function insert(el, container) {
  container.append(el);
}

const renderer: any = createRenderer({
  createElement,
  patchProp,
  insert,
});

export function createApp(...args) {
  return renderer.createApp(...args);
}
