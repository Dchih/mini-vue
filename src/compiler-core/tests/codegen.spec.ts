import { generate } from "../src/codegen";
import { baseParser } from "../src/parser";
import { transform } from "../src/transform";
import { transformExpression } from "../src/transforms/transformExpression";

describe("generate code", () => {
  test("string", () => {
    const ast = baseParser("hi");
    transform(ast);
    const { code } = generate(ast);

    expect(code).toMatchSnapshot();
  });
  test("interpolation", () => {
    const ast = baseParser("{{message}}");
    transform(ast, {
      nodeTransforms: [transformExpression],
    });
    const { code } = generate(ast);

    expect(code).toMatchSnapshot();
  });
});
