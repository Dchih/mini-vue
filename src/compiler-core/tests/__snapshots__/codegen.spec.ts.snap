// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate code element 1`] = `
"const { createElementBlock:_createElementBlock } = Vue
return function render(_ctx, _cache) {return _createElementBlock('div', null, null)}"
`;

exports[`generate code interpolation 1`] = `
"const { toDisplayString:_toDisplayString } = Vue
return function render(_ctx, _cache) {return _toDisplayString(_ctx.message)}"
`;

exports[`generate code string 1`] = `
"
return function render(_ctx, _cache) {return 'hi'}"
`;

exports[`generate code union type 1`] = `
"const { toDisplayString:_toDisplayString, createElementBlock:_createElementBlock } = Vue
return function render(_ctx, _cache) {return _createElementBlock('div', null, 'hi,' + _toDisplayString(_ctx.message))}"
`;
